File: Python/Arrays and Hashing/ProductOfArrayExceptSelf.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = [1]*len(nums) # initialize list with a value
        prefix = 1
        suffix = 1
        n = len(nums)

        for i in range(n):
            ans[i] *= prefix;
            prefix *= nums[i];
            ans[n-i-1] *= suffix;
            suffix *= nums[n-i-1];

        return ans


#make sure you don't use space for storing prefix and suffix

File: C++/Arrays and Hashing/ProductOfArrayExceptSelf.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int prefixNum = 1, suffixNum = 1, n = nums.size();
        vector<int> ans(n,1);

        for(int i=0 ; i<n ; i++){
            ans[i] *= prefixNum;
            prefixNum *= nums[i];
            ans[n-i-1] *= suffixNum;
            suffixNum *= nums[n-i-1];
        }

        return ans;
    }
};